### ISSUE TITLE
VictoriaMetrics (single-node) v1.77.0 panics during some kubernetes discovery process 

### ISSUE THREAD

**Issue by vpedosyuk** [VictoriaMetrics/VictoriaMetrics#2645]
**Describe the bug**

The system was working fine when at some point we noticed that there's no telemetry on Grafana. It lasted ~15 minutes then the pod of VictoriaMetrics restarted and it worked fine again.

When I checked logs of the previous container (the failed one), I saw many errors that VM failed "to scrape targets":

```json
{"ts":"2022-05-25T10:30:56.018Z","level":"error","caller":"VictoriaMetrics/lib/promscrape/scrapework.go:355","msg":"error when scraping \"http://10.35.80.3:8481/metrics\" from job \"kubernetes-pods\" with labels {app=\"<redacted>\",app_kubernetes_io_instance=\"<redacted>\",app_kubernetes_io_managed_by=\"Helm\",app_kubernetes_io_name=\"<redacted>\",helm_sh_chart=\"<redacted>\",instance=\"10.35.80.3:8481\",job=\"kubernetes-pods\",kubernetes_namespace=\"<redacted>\",kubernetes_pod_name=\"<redacted>\",pod_template_hash=\"8697785649\"}: error when scraping \"http://10.35.80.3:8481/metrics\": dial tcp4 10.35.80.3:8481: connect: connection refused; try -enableTCP6 command-line flag if you scrape ipv6 addresses"}
```

```json
{"ts":"2022-05-25T10:33:29.228Z","level":"error","caller":"VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:646","msg":"error when reading WatchEvent stream from \"https://10.35.240.1:443/api/v1/services?watch=1&allowWatchBookmarks=true&timeoutSeconds=1620&resourceVersion=1119611127\": unexpected EOF"}
```

```json
{"ts":"2022-05-25T10:37:34.350Z","level":"error","caller":"VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:534","msg":"cannot perform request to \"https://10.35.240.1:443/api/v1/endpoints\": Get \"https://10.35.240.1:443/api/v1/endpoints\": dial tcp 10.35.240.1:443: connect: connection refused"}
```

```json
{"ts":"2022-05-25T10:37:41.774Z","level":"error","caller":"VictoriaMetrics/lib/promscrape/scrapework.go:355","msg":"error when scraping \"https://kubernetes.default.svc:443/api/v1/nodes/<redacted>/proxy/metrics/cadvisor\" from job \"kubernetes-nodes-cadvisor\" with labels {beta_kubernetes_io_arch=\"amd64\",beta_kubernetes_io_instance_type=\"<redacted>\",beta_kubernetes_io_os=\"linux\",cloud_google_com_gke_boot_disk=\"<redacted>\",cloud_google_com_gke_container_runtime=\"<redacted>\",cloud_google_com_gke_nodepool=\"<redacted>\",cloud_google_com_gke_os_distribution=\"<redacted>\",cloud_google_com_gke_preemptible=\"true\",cloud_google_com_machine_family=\"<redacted>\",failure_domain_beta_kubernetes_io_region=\"<redacted>\",failure_domain_beta_kubernetes_io_zone=\"<redacted>\",instance=\"<redacted>\",job=\"kubernetes-nodes-cadvisor\",kubernetes_io_arch=\"amd64\",kubernetes_io_hostname=\"<redacted>\",kubernetes_io_os=\"linux\",node_kubernetes_io_instance_type=\"<redacted>\",topology_kubernetes_io_region=\"<redacted>\",topology_kubernetes_io_zone=\"<redacted>\"}: unexpected status code returned when scraping \"https://kubernetes.default.svc:443/api/v1/nodes/<redacted>/proxy/metrics/cadvisor\": 503; expecting 200; response body: \"{\\\"kind\\\":\\\"Status\\\",\\\"apiVersion\\\":\\\"v1\\\",\\\"metadata\\\":{},\\\"status\\\":\\\"Failure\\\",\\\"message\\\":\\\"error trying to reach service: rpc error: code = Unavailable desc = connection error: desc = \\\\\\\"transport: Error while dialing dial unix /etc/srv/kubernetes/konnectivity-server/konnectivity-server.socket: connect: no such file or directory\\\\\\\"\\\",\\\"reason\\\":\\\"ServiceUnavailable\\\",\\\"code\\\":503}\\n\""}
```

And in the end:
```
panic: runtime error: makeslice: cap out of range

goroutine 435 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/discovery/kubernetes.(*apiWatcher).getScrapeWorkObjects(0xc00432b8c0)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:155 +0x65
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/discovery/kubernetes.(*SDConfig).GetScrapeWorkObjects(...)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/discovery/kubernetes/kubernetes.go:78
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*Config).getKubernetesSDScrapeWork(0xc00487ad20, {0xc005c4a000, 0x1f5, 0x4aec77?})
	github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/config.go:723 +0x531
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.runScraper.func11(0xc0049e4680?, {0xc005c4a000?, 0xc006225920?, 0xc0047ca8a0?})
	github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/scraper.go:123 +0x1e
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scrapeConfig).run.func1(0xc00a0e7f50?)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/scraper.go:256 +0x86
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scrapeConfig).run(0xc0047d4340, 0x657079745f65636e?)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/scraper.go:272 +0x1b5
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scrapeConfigs).add.func1()
	github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/scraper.go:213 +0x66
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scrapeConfigs).add
	github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/scraper.go:211 +0x219
```

We've faced the issue a few times already. It happens once in a week or two.

**Expected behavior**
The issue wouldn't happen at all.

**Screenshots**
<img src="https://user-images.githubusercontent.com/16594968/170456971-e784cdbd-60e5-4339-a3dd-613400c40215.png" width="200"/>

**Version**
v1.77.0

**Used command-line flags**
We use Helm v0.8.27 with the following config:
```yaml
server:
  extraArgs:
    maxLabelsPerTimeseries: 50
    vmalert.proxyURL: http://<redacted>:8880
  scrape:
    enabled: true
    config:
      global:
        scrape_interval: 15s
        scrape_timeout: 15s
```

**Comment by f41gh7** [comment]
Hello, can you try v1.72.2 version? There was fix for kubernetes_sd

https://github.com/VictoriaMetrics/VictoriaMetrics/commit/d5eb6afe262dbbc392b50657b9f2d32dd3fc3cbe

**Comment by vpedosyuk** [comment]
@f41gh7, thanks, I'll try and report back if it reproduces again.

**Comment by valyala** [comment]
Closing the issue as fixed in [v1.77.2](https://docs.victoriametrics.com/CHANGELOG.html#v1772). @vpedosyuk , feel free re-opening the issue if it still exists in newer releases of VictoriaMetrics.
